JotunnLib/Managers/DungeonManager.cs

```diff
@@ -0,0 +1,387 @@
+        ///     The singleton instance of this manager.
+        /// </summary>
+        public static DungeonManager Instance => _instance ??= new DungeonManager();
+
+        /// <summary>
+        ///     Event that gets fired after the vanilla <see cref="DungeonDB"/> has loaded rooms.  Your code will execute
+        ///     every time a main scene is started (on joining a game). <br />
+        ///     If you want to execute just once you will need to unregister from the event after execution.
+        /// </summary>
+        public static event Action OnVanillaRoomsAvailable;
+
+        /// <summary>
+        ///     Internal dictionary of all custom rooms
+        /// </summary>
+        internal readonly Dictionary<string, CustomRoom> Rooms = new Dictionary<string, CustomRoom>();
+
+        private readonly Dictionary<int, string> hashToName = new Dictionary<int, string>();
+        private readonly List<string> themeList = new List<string>();
+        private readonly Dictionary<string, AssetBundle> prefabNames = new Dictionary<string, AssetBundle>();
+        private readonly Dictionary<string, AssetBundle> envNames = new Dictionary<string, AssetBundle>();
+
+        /// <summary>
+        ///     Container for Jötunn's DungeonRooms in the DontDestroyOnLoad scene.
+        /// </summary>
+        internal GameObject DungeonRoomContainer;
+
+
+        /// <summary>
+        ///     Hide .ctor
+        /// </summary>
+        private DungeonManager() { }
+
+        static DungeonManager()
+        {
+            ((IManager)Instance).Init();
+        }
+
+
+        /// <summary>
+        ///     Creates the spawner container and registers all hooks.
+        /// </summary>
+        void IManager.Init()
+        {
+            Main.LogInit("DungeonManager");
+
+            DungeonRoomContainer = new GameObject("DungeonRooms");
+            DungeonRoomContainer.transform.parent = Main.RootObject.transform;
+            DungeonRoomContainer.SetActive(false);
+
+            Main.Harmony.PatchAll(typeof(Patches));
+        }
+
+        private static class Patches
+        {
+            [HarmonyPatch(typeof(ZoneSystem), nameof(ZoneSystem.SetupLocations)), HarmonyPrefix]
+            private static void OnBeforeZoneSystemSetupLocations() => Instance.OnZoneSystemSetupLocations();
+
+            [HarmonyPatch(typeof(DungeonDB), nameof(DungeonDB.Start)), HarmonyPostfix]
+            private static void OnDungeonDBStarted() => Instance.OnDungeonDBStarted();
+
+            [HarmonyPatch(typeof(DungeonDB), nameof(DungeonDB.GetRoom)), HarmonyPrefix]
+            private static bool OnDungeonDBGetRoom(int hash, ref DungeonDB.RoomData __result)
+            {
+                DungeonDB.RoomData result = Instance.OnDungeonDBGetRoom(hash);
+
+                if (result != null)
+                {
+                    __result = result;
+                    return false;
+                }
+                return true;
+            }
```

Comment:
The room is already `DungeonDB.instance.m_rooms` and the `DungeonDB.instance.m_roomByHash` is regenerated, so all this extra handling doesn't look necessary

Reply:
The extra handling is necessary because `DungeonDB.instance.m_rooms` and `DungeonDB.instance.m_roomByHash` both do not contain any rooms with a Theme value of 0.  By default DungeonGenerator omits any rooms marked with either Theme=0 or Enabled=False.  Rather than patching the Generator to change the necessary logic (potential future maintenance headaches), it keeps it's own list of custom rooms and patches both `DungeonGenerator.SetupAvailableRooms` and `DungeonDB.GetRoom` to ensure the custom rooms are available for fully custom dungeons or for custom rooms for vanilla dungeons.

Reply:
I see, missed the theme handling.
However, it seems undesirable to have `DungeonDB.GetRooms()` and `DungeonDB.GetRoom(hash)` behave differently. In my opinion, it would be cleaner to either adding the rooms to the vanilla system and patching `SetupAvailableRooms` or not adding them to the vanilla system at all and only patching all necessary methods.

Reply:
Yeah that argument makes sense.  I think there was more of a necessity for the blended approach back when DungeonDB did more setup, before they added the soft reference system.

I'll pull out the logic related to and using `RegisterRoomInDungeonDB`, and see if there are any functional gaps that need to be patched.

Reply:
@MSchmoecker I'm also struggling with this comment. What am I supposed to change here?

Reply:
There's no clear suggestion to make here. It should work and behave consistently if possible, ie. the current implementation works but isn't consistent. If you don't have a better way, you can leave it as it is

Reply:
maybe the room should just be added to `DungeonDB.m_rooms` and `DungeonDB.m_roomByHash`. Then there isn't a need for the patch and both methods behave consistently